### PR TITLE
feat(provisioning): partnership user UI restrictions

### PR DESCRIPTION
**Pull request by cathteng**
Requires #59222 (backend)

Prevent partnership users from:
- Leaving the organization
- Being removed from the organization
- Having their role changed within the organization

Cannot remove user or change role

https://github.com/getsentry/sentry/assets/70817427/7f83fbb0-2477-4205-af50-71224c7cf11d

Cannot leave org

https://github.com/getsentry/sentry/assets/70817427/8e650d08-70e1-47b9-ab76-0f2f5088c8d4

For ER-1962

### CODE REVIEW

File: static/app/types/organization.tsx [hunk 1]
```diff
@@ -132,6 +132,7 @@ export interface Member {
     'member-limit:restricted': boolean;
     'sso:invalid': boolean;
     'sso:linked': boolean;
+    'partnership:restricted'?: boolean;
```
Comment (cathteng) [inline static/app/types/organization.tsx:135]: there are getsentry typescript and lint tests that fail if this is required

Comment (cathteng) [inline static/app/types/organization.tsx:135]: there are getsentry typescript and lint tests that fail if this is required -- will remove the `?` when the relevant files are updated

File: static/app/views/settings/organizationMembers/organizationMemberDetail.tsx [hunk 2]
```diff
@@ -286,6 +286,7 @@ class OrganizationMemberDetail extends DeprecatedAsyncView<Props, State> {
 
     const {access, orgRoleList} = organization;
     const canEdit = access.includes('org:write') && !this.memberDeactivated;
+    const isPartnershipUser = !!member.flags['partnership:restricted'];
```
Comment (cathteng) [inline static/app/views/settings/organizationMembers/organizationMemberDetail.tsx:289]: this casts the undefined to false if necessary

File: static/app/views/settings/organizationMembers/organizationMemberDetail.tsx [hunk 3]
```diff
@@ -383,10 +384,11 @@ class OrganizationMemberDetail extends DeprecatedAsyncView<Props, State> {
         <OrganizationRoleSelect
           enforceAllowed={false}
           enforceRetired={this.hasTeamRoles}
-          disabled={!canEdit}
+          disabled={!canEdit || isPartnershipUser}
           roleList={orgRoleList}
           roleSelected={orgRole}
           setSelected={this.onChangeOrgRole}
+          helpText={isPartnershipUser ? t('This is a partnership user.') : undefined}
```
Comment (cathteng) [inline static/app/views/settings/organizationMembers/organizationMemberDetail.tsx:391]: i need some text suggestions that aren't too long

File: static/app/views/settings/organizationMembers/organizationMemberDetail.tsx (hunk 2 above)
Comment (sentaur-athena) [reply static/app/views/settings/organizationMembers/organizationMemberDetail.tsx:289]: You can do `member.flags['partnership:restricted'] === true` to be more clear.

File: static/app/views/settings/organizationMembers/organizationMemberDetail.tsx (hunk 3 above)
Comment (sentaur-athena) [reply static/app/views/settings/organizationMembers/organizationMemberDetail.tsx:391]: cc: @Dhrumil-Sentry 
I'm really bad at this. How about `Partnership users can't be modified`?

File: static/app/views/settings/organizationMembers/organizationMemberRow.tsx [hunk 4]
```diff
@@ -230,6 +234,8 @@ export default class OrganizationMemberRow extends PureComponent<Props, State> {
                     ? t(
                         "Your account is managed through your organization's identity provider."
                       )
+                    : isPartnershipUser
+                    ? t('You are a partnership user.')
```
Comment (sentaur-athena) [inline static/app/views/settings/organizationMembers/organizationMemberRow.tsx:238]: We might be able to use the same message here. Like `Partnership users can't be modified` or whatever.

Comment (cathteng) [reply static/app/views/settings/organizationMembers/organizationMemberRow.tsx:238]: yep i would use the same message everywhere

File: static/app/views/settings/organizationMembers/organizationMemberDetail.tsx (hunk 3 above)
Comment (schew2381) [reply static/app/views/settings/organizationMembers/organizationMemberDetail.tsx:391]: 1. `Cannot change the role of partnership users`
2. `It's not possible to alter the partnership users' roles.`
3. `Modifying the roles of partnership users is not an option.`
4. `Partnership users' roles are fixed and cannot be changed.`
5. `The roles of partnership users are unmodifiable.`
6. `There is no way to adjust the partnership users' roles.`

File: static/app/views/settings/organizationMembers/organizationMemberRow.tsx [hunk 5]
```diff
@@ -199,6 +201,8 @@ export default class OrganizationMemberRow extends PureComponent<Props, State> {
                     ? t(
                         "This user is managed through your organization's identity provider."
                       )
+                    : isPartnershipUser
+                    ? t('This is a partnership user.')
```
Comment (schew2381) [inline static/app/views/settings/organizationMembers/organizationMemberRow.tsx:205]: `Partnership Users cannot be removed{ from the organization? (maybe)}`

**Comment by Dhrumil-Sentry** [comment]
@cathteng @sentaur-athena can we do "You cannot make changes to this partner provisioned account"

File: static/app/views/settings/organizationMembers/organizationMemberRow.tsx (hunk 4 above)
Comment (schew2381) [inline static/app/views/settings/organizationMembers/organizationMemberRow.tsx:238]: If the other string is updated, this can maybe be the same

**Review by sentaur-athena — APPROVED** [review]
Approving to unblock. You have some suggestions for the message now.

**Pull request merged**